File: internal/config/observability.go
//
// Copyright (C) 2019-2026 vdaas.org vald team <vald@vdaas.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package config

// Observability represents the configuration for the observability.
type Observability struct {
	// OTLP represents the OTLP configuration.
	OTLP *OTLP `json:"otlp" yaml:"otlp"`
	// Metrics represents the metrics configuration.
	Metrics *Metrics `json:"metrics" yaml:"metrics"`
	// Trace represents the trace configuration.
	Trace *Trace `json:"trace" yaml:"trace"`
	// Enabled enables observability.
	Enabled bool `json:"enabled" yaml:"enabled"`
}

// Bind binds the actual data from the Observability receiver fields.
func (o *Observability) Bind() *Observability {
	if o.OTLP == nil {
		o.OTLP = new(OTLP)
	}
	if o.OTLP != nil {
		o.OTLP.Bind()
	}

	if o.Metrics == nil {
		o.Metrics = new(Metrics)
	}
	if o.Metrics != nil {
		o.Metrics.Bind()
	}

	if o.Trace == nil {
		o.Trace = new(Trace)
	}
	if o.Trace != nil {
		o.Trace.Bind()
	}

	return o
}

// OTLP represents the configuration for the OTLP.
type OTLP struct {
	// Attribute represents the OTLP attribute configuration.
	Attribute *OTLPAttribute `json:"attribute" yaml:"attribute"`
	// CollectorEndpoint represents the collector endpoint.
	CollectorEndpoint string `json:"collector_endpoint" yaml:"collector_endpoint"`
	// TraceBatchTimeout represents the trace batch timeout.
	TraceBatchTimeout string `json:"trace_batch_timeout" yaml:"trace_batch_timeout"`
	// TraceExportTimeout represents the trace export timeout.
	TraceExportTimeout string `json:"trace_export_timeout" yaml:"trace_export_timeout"`
	// MetricsExportInterval represents the metrics export interval.
	MetricsExportInterval string `json:"metrics_export_interval" yaml:"metrics_export_interval"`
	// MetricsExportTimeout represents the metrics export timeout.
	MetricsExportTimeout string `json:"metrics_export_timeout" yaml:"metrics_export_timeout"`
	// TraceMaxExportBatchSize represents the trace max export batch size.
	TraceMaxExportBatchSize int `json:"trace_max_export_batch_size" yaml:"trace_max_export_batch_size"`
	// TraceMaxQueueSize represents the trace max queue size.
	TraceMaxQueueSize int `json:"trace_max_queue_size" yaml:"trace_max_queue_size"`
}

// Bind binds the actual data from the OTLP receiver fields.
func (o *OTLP) Bind() *OTLP {
	o.CollectorEndpoint = GetActualValue(o.CollectorEndpoint)
	o.TraceBatchTimeout = GetActualValue(o.TraceBatchTimeout)
	o.TraceExportTimeout = GetActualValue(o.TraceExportTimeout)
	o.MetricsExportInterval = GetActualValue(o.MetricsExportInterval)
	o.MetricsExportTimeout = GetActualValue(o.MetricsExportTimeout)
	if o.Attribute == nil {
		o.Attribute = new(OTLPAttribute)
	}
	if o.Attribute != nil {
		o.Attribute.Bind()
	}
	return o
}

// OTLPAttribute represents the configuration for the OTLP attribute.
type OTLPAttribute struct {
	// Namespace represents the namespace.
	Namespace string `json:"namespace" yaml:"namespace"`
	// PodName represents the pod name.
	PodName string `json:"pod_name" yaml:"pod_name"`
	// NodeName represents the node name.
	NodeName string `json:"node_name" yaml:"node_name"`
	// ServiceName represents the service name.
	ServiceName string `json:"service_name" yaml:"service_name"`
}

// Bind binds the actual data from the OTLPAttribute receiver fields.
func (o *OTLPAttribute) Bind() *OTLPAttribute {
	o.Namespace = GetActualValue(o.Namespace)
	o.PodName = GetActualValue(o.PodName)
	o.NodeName = GetActualValue(o.NodeName)
	o.ServiceName = GetActualValue(o.ServiceName)
	return o
}

// Trace represents the configuration for the trace.
type Trace struct {
	// Enabled enables trace.
	Enabled bool `json:"enabled" yaml:"enabled"`
}

// Bind binds the actual data from the Trace receiver fields.
func (t *Trace) Bind() *Trace {
	// No fields to bind as per rules
	return t
}

// Metrics represents the configuration for the metrics.
type Metrics struct {
	// VersionInfoLabels represents the version info labels.
	VersionInfoLabels []string `json:"version_info_labels" yaml:"version_info_labels"`
	// EnableVersionInfo enables version info metrics.
	EnableVersionInfo bool `json:"enable_version_info" yaml:"enable_version_info"`
	// EnableMemory enables memory metrics.
	EnableMemory bool `json:"enable_memory" yaml:"enable_memory"`
	// EnableGoroutine enables goroutine metrics.
	EnableGoroutine bool `json:"enable_goroutine" yaml:"enable_goroutine"`
	// EnableCGO enables cgo metrics.
	EnableCGO bool `json:"enable_cgo" yaml:"enable_cgo"`
}

// Bind binds the actual data from the Metrics receiver fields.
func (m *Metrics) Bind() *Metrics {
	m.VersionInfoLabels = GetActualValues(m.VersionInfoLabels)
	return m
}

File: rust/libs/observability/src/observability.rs
//
// Copyright (C) 2019-2026 vdaas.org vald team <vald@vdaas.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
use anyhow::{Ok, Result};
use opentelemetry::global;
use opentelemetry_otlp::{MetricExporter, SpanExporter, WithExportConfig};
use opentelemetry_sdk::metrics::{PeriodicReader, SdkMeterProvider};
use opentelemetry_sdk::propagation::TraceContextPropagator;
use opentelemetry_sdk::trace::{self, SdkTracerProvider};
use opentelemetry_sdk::Resource;
use url::Url;

use crate::config::Config;

pub const SERVICE_NAME: &str = opentelemetry_semantic_conventions::resource::SERVICE_NAME;

pub trait Observability {
    fn shutdown(&mut self) -> Result<()>;
}

pub struct ObservabilityImpl {
    config: Config,
    meter_provider: Option<SdkMeterProvider>,
    tracer_provider: Option<SdkTracerProvider>,
}

impl ObservabilityImpl {
    pub fn new(cfg: Config) -> Result<ObservabilityImpl, anyhow::Error> {
        let mut obj = ObservabilityImpl {
            config: cfg,
            meter_provider: None,
            tracer_provider: None,
        };

        if !obj.config.enabled {
            return Ok(obj);
        }

        if obj.config.meter.enabled {
            // NOTE: Since the agent implementation does not use views, we will use the simplest implementation for the current phase.
            // If we want flexibility and customization, use SdkMeterProvider::builder.
            let exporter = MetricExporter::builder()
                .with_tonic()
                .with_endpoint(
                    Url::parse(obj.config.endpoint.as_str())?
                        .join("/v1/metrics")?
                        .as_str(),
                )
                .with_timeout(obj.config.meter.export_timeout_duration)
                .build()?;
            let reader = PeriodicReader::builder(exporter)
                .with_interval(obj.config.meter.export_duration)
                .build();
            let provider = SdkMeterProvider::builder()
                .with_reader(reader)
                .with_resource(Resource::from(obj.config()))
                .build();
            obj.meter_provider = Some(provider.clone());
            global::set_meter_provider(provider.clone());
        }

        if obj.config.tracer.enabled {
            let exporter = SpanExporter::builder()
                .with_tonic()
                .with_endpoint(
                    Url::parse(obj.config.endpoint.as_str())?
                        .join("/v1/traces")?
                        .as_str(),
                )
                .build()?;
            let provider = SdkTracerProvider::builder()
                .with_batch_exporter(exporter)
                .with_sampler(trace::Sampler::AlwaysOn)
                .with_resource(Resource::from(obj.config()))
                .with_id_generator(trace::RandomIdGenerator::default())
                .build();
            obj.tracer_provider = Some(provider.clone());
            global::set_text_map_propagator(TraceContextPropagator::new());
            global::set_tracer_provider(provider.clone());
        }
        Ok(obj)
    }

    fn config(&self) -> &Config {
        &self.config
    }
}

impl Observability for ObservabilityImpl {
    fn shutdown(&mut self) -> Result<()> {
        if !self.config.enabled {
            return Ok(());
        }

        if self.config.meter.enabled {
            if let Some(ref provider) = self.meter_provider {
                provider.force_flush()?;
                provider.shutdown()?;
            }
        }

        if self.config.tracer.enabled {
            if let Some(ref provider) = self.tracer_provider {
                provider.force_flush()?;
                provider.shutdown()?;
            }
        }
        Ok(())
    }
}
